javacutil/src/main/java/org/checkerframework/javacutil/SwitchExpressionScanner.java

```diff
@@ -0,0 +1,171 @@
+   * does not descend into switch expressions.
+   */
+  protected YieldVisitor yieldVisitor = new YieldVisitor();
+
+  /**
+   * A scanner that visits all the yield trees in a given tree and calls {@link
+   * #visitSwitchResultExpression(ExpressionTree, Object)} on the expression in the yield trees. It
+   * does not descend into switch expressions.
+   */
+  protected class YieldVisitor extends TreeScanner<@Nullable R, P> {
+
+    @Override
+    public @Nullable R scan(Tree tree, P p) {
+      if (tree == null) {
+        return null;
+      }
+      if (tree.getKind().name().equals("SWITCH_EXPRESSION")) {
+        // Don't scan nested switch expressions.
+        return null;
+      } else if (tree.getKind().name().equals("YIELD")) {
+        ExpressionTree value = TreeUtils.yieldTreeGetValue(tree);
+        return visitSwitchResultExpression(value, p);
+      }
+      return super.scan(tree, p);
+    }
+
+    @Override
+    public R reduce(R r1, R r2) {
+      return combineResults(r1, r2);
+    }
+  }
+
+  /**
+   * An implementation of {@link SwitchExpressionScanner} that uses functions {@link
+   * #visitSwitchResultExpression(ExpressionTree, Object)} and {@link #combineResults(Object,
+   * Object)}.
+   *
+   * @param <R1> the type result of {@link #visitSwitchResultExpression(ExpressionTree, Object)}
+   * @param <P1> the type of the parameter to pass to {@link
+   *     #visitSwitchResultExpression(ExpressionTree, Object)}
+   */
+  public static class FunctionalSwitchExpressionScanner<R1, P1>
+      extends SwitchExpressionScanner<R1, P1> {
+
+    /** The function to use for {@link #visitSwitchResultExpression(ExpressionTree, Object)}. */
+    private final BiFunction<ExpressionTree, P1, R1> switchValueExpressionFunction;
+    /** The function to use for {@link #visitSwitchResultExpression(ExpressionTree, Object)}. */
+    private final BiFunction<@Nullable R1, @Nullable R1, R1> combineResultFunc;
+
+    /**
+     * Creates a {@link FunctionalSwitchExpressionScanner} that uses the given functions.
+     *
+     * @param switchValueExpressionFunc the function called on each switch result expression
+     * @param combineResultFunc the function used to combine the result of multiple calls to {@code
+     *     switchValueExpressionFunc}
+     */
+    public FunctionalSwitchExpressionScanner(
+        BiFunction<ExpressionTree, P1, R1> switchValueExpressionFunc,
+        BiFunction<@Nullable R1, @Nullable R1, R1> combineResultFunc) {
+      this.switchValueExpressionFunction = switchValueExpressionFunc;
+      this.combineResultFunc = combineResultFunc;
+    }
+
+    @Override
+    protected R1 visitSwitchResultExpression(ExpressionTree resultExpressionTree, P1 p1) {
+      return switchValueExpressionFunction.apply(resultExpressionTree, p1);
+    }
+
+    @Override
+    protected R1 combineResults(@Nullable R1 r1, @Nullable R1 r2) {
+      return combineResultFunc.apply(r1, r2);
```

Comment:
There could be cases where null should be handled differently.  But, I didn't special case null so that the specification for `FunctionalSwitchExpressionScanner#combineResultFunc` is identical to `SwitchExpressionScanner#combineResults`.